File: packages/typescript/src/agent/agent.ts
/**
 * Core Agent class — the central abstraction of the Scalix SDK.
 */

import { getConfig, isCloudMode, dynamicImport } from '../config.js';
import type { Tool } from '../tools/base.js';
import type {
  AgentOptions,
  AgentResult,
  Message,
  StreamEvent,
  ToolCall,
  ToolCallResult,
  ToolDefinition,
} from '../types.js';
import type { LLMProvider } from '../providers/base.js';

/**
 * An AI agent that can reason, use tools, and maintain state.
 *
 * The agent runs an LLM in a loop: it sends messages to the model,
 * executes any tool calls the model requests, feeds results back,
 * and repeats until the model produces a final text response.
 *
 * @example
 * ```typescript
 * import { Agent, Tool } from 'scalix';
 *
 * const agent = new Agent({
 *   model: 'claude-sonnet-4',
 *   instructions: 'You are a helpful data analyst.',
 *   tools: [Tool.codeExec(), Tool.webSearch()],
 * });
 *
 * const result = await agent.run('Analyze the top trending repos');
 * console.log(result.output);
 * ```
 */
export class Agent {
  readonly model: string;
  readonly instructions: string;
  readonly tools: Tool[];
  readonly memory: boolean;
  readonly maxTurns: number;
  readonly temperature: number;
  readonly timeout: number;

  private history: Message[] = [];
  private llmProvider: LLMProvider | null = null;
  private toolExecutor: ToolExecutor | null = null;

  constructor(options: AgentOptions = {}) {
    this.model = options.model ?? 'auto';
    this.instructions = options.instructions ?? '';
    this.tools = options.tools ?? [];
    this.memory = options.memory ?? false;
    this.maxTurns = options.maxTurns ?? 10;
    this.temperature = options.temperature ?? 0.7;
    this.timeout = options.timeout ?? 300;
  }

  private async getLLMProvider(): Promise<LLMProvider> {
    if (!this.llmProvider) {
      if (isCloudMode()) {
        const { ScalixRouterProvider } = await import('../providers/cloud/router.js');
        this.llmProvider = new ScalixRouterProvider(getConfig()) as LLMProvider;
      } else {
        const { DirectLLM } = await import('../providers/local/directLlm.js');
        this.llmProvider = new DirectLLM(getConfig()) as LLMProvider;
      }
    }
    return this.llmProvider;
  }

  private getToolExecutor(): ToolExecutor {
    if (!this.toolExecutor) {
      this.toolExecutor = new ToolExecutor(this.tools, getConfig());
    }
    return this.toolExecutor;
  }

  /**
   * Execute the agent with the given prompt.
   *
   * The agent will call the LLM, execute any tool calls, and loop
   * until it produces a final text response or hits maxTurns.
   */
  async run(prompt: string): Promise<AgentResult> {
    const llm = await this.getLLMProvider();
    const executor = this.getToolExecutor();

    // Build initial messages
    const messages = this.buildMessages(prompt);
    const allToolResults: ToolCallResult[] = [];

    // Get tool definitions for the LLM
    const toolDefs = this.tools.length > 0
      ? this.tools.map((t) => t.definition as unknown as Record<string, unknown>)
      : undefined;

    let hitMaxTurns = false;

    for (let turn = 0; turn < this.maxTurns; turn++) {
      // Call LLM
      const response = await llm.chat(messages, this.model, {
        tools: toolDefs,
        temperature: this.temperature,
      });

      messages.push(response);

      // If no tool calls, we have the final response
      if (!response.toolCalls?.length) {
        break;
      }

      // Execute each tool call
      for (const toolCall of response.toolCalls) {
        const toolResult = await executor.execute(toolCall);
        allToolResults.push(toolResult);

        // Add tool result as a message for the next LLM call
        let resultContent: string;
        if (toolResult.error) {
          resultContent = `Error: ${toolResult.error}`;
        } else if (typeof toolResult.result === 'string') {
          resultContent = toolResult.result;
        } else {
          resultContent = JSON.stringify(toolResult.result);
        }

        messages.push({
          role: 'tool',
          content: resultContent,
          toolCallId: toolCall.id,
          name: toolCall.name,
        });
      }

      // Check if this was the last turn
      if (turn === this.maxTurns - 1) {
        hitMaxTurns = true;
      }
    }

    // If we hit max turns, get a final response without tools
    if (hitMaxTurns) {
      const finalResponse = await llm.chat(messages, this.model, {
        temperature: this.temperature,
      });
      messages.push(finalResponse);
    }

    // Update memory
    if (this.memory) {
      this.history = [...messages];
    }

    // Extract final output
    const finalOutput = messages.length > 0 ? messages[messages.length - 1].content : '';

    return {
      output: finalOutput,
      messages,
      toolCalls: allToolResults,
    };
  }

  /**
   * Execute the agent and stream events as they occur.
   */
  async *stream(prompt: string): AsyncGenerator<StreamEvent> {
    const llm = await this.getLLMProvider();
    const executor = this.getToolExecutor();
    const messages = this.buildMessages(prompt);

    const toolDefs = this.tools.length > 0
      ? this.tools.map((t) => t.definition as unknown as Record<string, unknown>)
      : undefined;

    for (let turn = 0; turn < this.maxTurns; turn++) {
      // If no tools, stream directly
      if (!this.tools.length) {
        yield* llm.streamChat(messages, this.model, {
          temperature: this.temperature,
        });
        return;
      }

      // With tools, get full response to check for tool calls
      const response = await llm.chat(messages, this.model, {
        tools: toolDefs,
        temperature: this.temperature,
      });

      messages.push(response);

      if (!response.toolCalls?.length) {
        // Stream the final text response
        yield { type: 'text_delta', data: response.content };
        yield { type: 'done' };
        return;
      }

      // Execute tools and stream their results
      for (const toolCall of response.toolCalls) {
        yield {
          type: 'tool_call_start',
          toolName: toolCall.name,
          toolCallId: toolCall.id,
          data: toolCall.arguments as unknown as Record<string, unknown>,
        };

        const toolResult = await executor.execute(toolCall);

        let resultContent: string;
        if (toolResult.error) {
          resultContent = `Error: ${toolResult.error}`;
        } else if (typeof toolResult.result === 'string') {
          resultContent = toolResult.result;
        } else {
          resultContent = JSON.stringify(toolResult.result);
        }

        yield {
          type: 'tool_result',
          toolName: toolCall.name,
          toolCallId: toolCall.id,
          data: resultContent,
        };

        messages.push({
          role: 'tool',
          content: resultContent,
          toolCallId: toolCall.id,
          name: toolCall.name,
        });
      }
    }

    yield { type: 'done' };
  }

  /**
   * Build the message list for an LLM call.
   */
  private buildMessages(prompt: string): Message[] {
    const messages: Message[] = [];

    // System prompt
    if (this.instructions) {
      messages.push({ role: 'system', content: this.instructions });
    }

    // History (if memory enabled)
    if (this.memory && this.history.length > 0) {
      for (const msg of this.history) {
        if (msg.role !== 'system') {
          messages.push(msg);
        }
      }
    }

    // User prompt
    messages.push({ role: 'user', content: prompt });

    return messages;
  }

  /**
   * Get JSON Schema definitions for all registered tools.
   */
  getToolDefinitions(): ToolDefinition[] {
    return this.tools.map((tool) => tool.definition);
  }

  /**
   * Clear the agent's conversation history.
   */
  clearMemory(): void {
    this.history = [];
  }
}

/**
 * Executes tool calls by routing to the appropriate handler.
 *
 * Maps tool names to their execution backends:
 * - code_exec → subprocess / Docker
 * - sql → better-sqlite3
 * - web_search → fetch-based search
 * - http → fetch request
 * - Custom tools → their registered execute functions
 */
class ToolExecutor {
  private tools: Map<string, Tool>;

  constructor(
    tools: Tool[],
    _config: ReturnType<typeof getConfig>,
  ) {
    this.tools = new Map(tools.map((t) => [t.name, t]));
  }

  async execute(toolCall: ToolCall): Promise<ToolCallResult> {
    const start = performance.now();

    const tool = this.tools.get(toolCall.name);
    if (!tool) {
      return {
        toolName: toolCall.name,
        arguments: toolCall.arguments,
        result: null,
        error: `Unknown tool: ${toolCall.name}`,
      };
    }

    try {
      let result: unknown;

      switch (toolCall.name) {
        case 'code_exec':
          result = await this.executeCode(toolCall.arguments);
          break;
        case 'sql':
          result = await this.executeSql(toolCall.arguments);
          break;
        case 'web_search':
          result = await this.executeWebSearch(toolCall.arguments);
          break;
        case 'http':
          result = await this.executeHttp(toolCall.arguments);
          break;
        default:
          // Custom tool with registered function
          result = await tool.execute(toolCall.arguments);
          return result as ToolCallResult;
      }

      const durationMs = performance.now() - start;
      return {
        toolName: toolCall.name,
        arguments: toolCall.arguments,
        result,
        durationMs,
      };
    } catch (err) {
      const durationMs = performance.now() - start;
      return {
        toolName: toolCall.name,
        arguments: toolCall.arguments,
        result: null,
        error: err instanceof Error ? err.message : String(err),
        durationMs,
      };
    }
  }

  private async executeCode(
    args: Record<string, unknown>,
  ): Promise<string> {
    const { execFile } = await import('child_process');
    const { promisify } = await import('util');
    const { writeFile, unlink } = await import('fs/promises');
    const { tmpdir } = await import('os');
    const { join } = await import('path');

    const execFileAsync = promisify(execFile);
    const code = (args.code as string) ?? '';
    const runtime = (args.runtime as string) ?? 'python';

    const runtimeCommands: Record<string, { cmd: string; flag: string }> = {
      python: { cmd: 'python3', flag: '-c' },
      node: { cmd: 'node', flag: '-e' },
    };

    const rtConfig = runtimeCommands[runtime];
    if (rtConfig) {
      // Inline execution for python/node
      try {
        const { stdout, stderr } = await execFileAsync(rtConfig.cmd, [rtConfig.flag, code], {
          timeout: 30_000,
        });
        let output = stdout;
        if (stderr) output += `\nSTDERR:\n${stderr}`;
        return output;
      } catch (err: unknown) {
        const e = err as { stdout?: string; stderr?: string; code?: number };
        let output = e.stdout ?? '';
        if (e.stderr) output += `\nSTDERR:\n${e.stderr}`;
        output += `\n[Exit code: ${e.code ?? 1}]`;
        return output;
      }
    }

    // File-based execution for go/rust
    const extMap: Record<string, string> = { go: '.go', rust: '.rs' };
    const ext = extMap[runtime] ?? '.txt';
    const filePath = join(tmpdir(), `scalix_${Date.now()}${ext}`);

    try {
      await writeFile(filePath, code, 'utf-8');
      const cmdMap: Record<string, string[]> = {
        go: ['go', 'run', filePath],
        rust: ['sh', '-c', `rustc ${filePath} -o /tmp/scalix_out && /tmp/scalix_out`],
      };
      const cmd = cmdMap[runtime];
      if (!cmd) throw new Error(`Unsupported runtime: ${runtime}`);

      const { stdout, stderr } = await execFileAsync(cmd[0], cmd.slice(1), {
        timeout: 30_000,
      });
      let output = stdout;
      if (stderr) output += `\nSTDERR:\n${stderr}`;
      return output;
    } finally {
      await unlink(filePath).catch(() => {});
    }
  }

  private async executeSql(args: Record<string, unknown>): Promise<unknown> {
    // Use better-sqlite3 for local mode (sync but fast)
    const query = (args.query as string) ?? '';
    try {
      const mod = await dynamicImport('better-sqlite3') as { default: new (path: string) => { prepare: (sql: string) => { all: () => unknown[]; run: () => { changes: number } }; close: () => void } };
      const db = new mod.default('scalix_local.db');
      const stmt = db.prepare(query);

      if (query.trim().toUpperCase().startsWith('SELECT')) {
        const rows = stmt.all();
        db.close();
        return rows;
      } else {
        const result = stmt.run();
        db.close();
        return { changes: result.changes };
      }
    } catch {
      throw new Error(
        'better-sqlite3 not installed. Run: npm install better-sqlite3',
      );
    }
  }

  private async executeWebSearch(args: Record<string, unknown>): Promise<string> {
    const query = (args.query as string) ?? '';
    const config = getConfig();
    const searchUrl = config.searchBaseUrl
      ?? `${config.baseUrl}/research/search`;
    const headers: Record<string, string> = {
      'User-Agent': 'Scalix-SDK/0.1',
      'Content-Type': 'application/json',
    };
    if (config.apiKey) headers['Authorization'] = `Bearer ${config.apiKey}`;
    const resp = await fetch(searchUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify({ query }),
    });
    const text = await resp.text();
    return text.slice(0, 5000);
  }

  private async executeHttp(args: Record<string, unknown>): Promise<unknown> {
    const url = (args.url as string) ?? '';
    const method = ((args.method as string) ?? 'GET').toUpperCase();
    const body = args.body as Record<string, unknown> | undefined;

    const resp = await fetch(url, {
      method,
      ...(body ? { body: JSON.stringify(body), headers: { 'Content-Type': 'application/json' } } : {}),
    });

    const text = await resp.text();
    return { status: resp.status, body: text.slice(0, 5000) };
  }
}

File: packages/python/scalix/agent/agent.py
"""Core Agent class — the central abstraction of the Scalix SDK."""

from __future__ import annotations

import json
import logging
import time
from typing import Any, AsyncIterator

from scalix.config import get_config, ScalixConfig
from scalix.exceptions import ProviderError, ToolError
from scalix.types import (
    AgentResult,
    Message,
    Role,
    StreamEvent,
    StreamEventType,
    ToolCall,
    ToolCallResult,
    ToolDefinition,
    Usage,
)

logger = logging.getLogger("scalix.agent")


class Agent:
    """An AI agent that can reason, use tools, and maintain state.

    The agent runs an LLM in a loop: it sends messages to the model,
    executes any tool calls the model requests, feeds results back,
    and repeats until the model produces a final text response.

    Args:
        model: LLM model to use. "auto" lets Scalix Router choose optimally.
        instructions: System prompt defining the agent's behavior.
        tools: List of tools the agent can use.
        memory: Whether to persist conversation history across runs.
        max_turns: Maximum number of tool-use loops before stopping.
        temperature: Sampling temperature for LLM responses.
        timeout: Maximum execution time in seconds.

    Example:
        from scalix import Agent, Tool

        agent = Agent(
            model="claude-sonnet-4",
            instructions="You are a helpful data analyst.",
            tools=[Tool.code_exec(), Tool.web_search()],
        )

        result = await agent.run("Analyze the top trending GitHub repos")
        print(result.output)
    """

    def __init__(
        self,
        model: str = "auto",
        instructions: str = "",
        tools: list[Any] | None = None,
        memory: bool = False,
        max_turns: int = 10,
        temperature: float = 0.7,
        timeout: int = 300,
    ) -> None:
        self.model = model
        self.instructions = instructions
        self.tools = tools or []
        self.memory = memory
        self.max_turns = max_turns
        self.temperature = temperature
        self.timeout = timeout

        self._history: list[Message] = []
        self._config = get_config()
        self._llm_provider: Any = None
        self._tool_executor: Any = None

    def _get_llm_provider(self) -> Any:
        """Get the LLM provider (lazy initialization)."""
        if self._llm_provider is None:
            if self._config.is_cloud_mode:
                from scalix.providers.cloud.router import ScalixRouterProvider

                self._llm_provider = ScalixRouterProvider(self._config)
            else:
                from scalix.providers.local.direct_llm import DirectLLM

                self._llm_provider = DirectLLM(self._config)
        return self._llm_provider

    def _get_tool_executor(self) -> ToolExecutor:
        """Get the tool executor (lazy initialization)."""
        if self._tool_executor is None:
            self._tool_executor = ToolExecutor(self.tools, self._config)
        return self._tool_executor

    async def run(self, prompt: str, **kwargs: Any) -> AgentResult:
        """Execute the agent with the given prompt.

        The agent will call the LLM, execute any tool calls, and loop
        until it produces a final text response or hits max_turns.

        Args:
            prompt: The user's input prompt.
            **kwargs: Additional parameters passed to the LLM.

        Returns:
            AgentResult with the agent's output, messages, and tool call history.
        """
        llm = self._get_llm_provider()
        executor = self._get_tool_executor()

        # Build initial messages
        messages = self._build_messages(prompt)
        all_tool_results: list[ToolCallResult] = []
        total_usage = Usage()
        used_model: str | None = None

        # Get tool definitions for the LLM
        tool_defs = [t.definition.model_dump() for t in self.tools] if self.tools else None

        for turn in range(self.max_turns):
            logger.debug("Agent turn %d/%d", turn + 1, self.max_turns)

            # Call LLM
            response = await llm.chat(
                messages=messages,
                model=self.model,
                tools=tool_defs,
                temperature=self.temperature,
                **kwargs,
            )

            messages.append(response)

            # If no tool calls, we have the final response
            if not response.tool_calls:
                logger.debug("Agent finished with text response")
                break

            # Execute each tool call
            for tool_call in response.tool_calls:
                logger.debug("Executing tool: %s", tool_call.name)
                tool_result = await executor.execute(tool_call)
                all_tool_results.append(tool_result)

                # Add tool result as a message for the next LLM call
                result_content = (
                    json.dumps(tool_result.result)
                    if not isinstance(tool_result.result, str)
                    else tool_result.result
                )
                if tool_result.error:
                    result_content = f"Error: {tool_result.error}"

                messages.append(
                    Message(
                        role=Role.TOOL,
                        content=result_content,
                        tool_call_id=tool_call.id,
                        name=tool_call.name,
                    )
                )
        else:
            # Hit max_turns — get final response without tools
            logger.warning("Agent hit max_turns (%d), forcing final response", self.max_turns)
            response = await llm.chat(
                messages=messages,
                model=self.model,
                temperature=self.temperature,
                **kwargs,
            )
            messages.append(response)

        # Update memory
        if self.memory:
            self._history = messages.copy()

        # Extract final output
        final_output = messages[-1].content if messages else ""

        return AgentResult(
            output=final_output,
            messages=messages,
            tool_calls=all_tool_results,
            model=used_model,
            usage=total_usage if total_usage.total_tokens > 0 else None,
        )

    async def stream(self, prompt: str, **kwargs: Any) -> AsyncIterator[StreamEvent]:
        """Execute the agent and stream events as they occur.

        Yields StreamEvent objects for text deltas, tool calls, and results.

        Args:
            prompt: The user's input prompt.
            **kwargs: Additional parameters passed to the LLM.

        Yields:
            StreamEvent objects representing each step of execution.
        """
        llm = self._get_llm_provider()
        executor = self._get_tool_executor()
        messages = self._build_messages(prompt)
        tool_defs = [t.definition.model_dump() for t in self.tools] if self.tools else None

        for turn in range(self.max_turns):
            # If no tools or non-streaming needed, fall back to run
            if not self.tools:
                async for event in llm.stream_chat(
                    messages=messages,
                    model=self.model,
                    temperature=self.temperature,
                    **kwargs,
                ):
                    yield event
                return

            # With tools, we need the full response to check for tool calls
            response = await llm.chat(
                messages=messages,
                model=self.model,
                tools=tool_defs,
                temperature=self.temperature,
                **kwargs,
            )

            messages.append(response)

            if not response.tool_calls:
                # Stream the final text response
                yield StreamEvent(type=StreamEventType.TEXT_DELTA, data=response.content)
                yield StreamEvent(type=StreamEventType.DONE)
                return

            # Execute tools and stream their results
            for tool_call in response.tool_calls:
                yield StreamEvent(
                    type=StreamEventType.TOOL_CALL_START,
                    tool_name=tool_call.name,
                    tool_call_id=tool_call.id,
                    data=tool_call.arguments,
                )

                tool_result = await executor.execute(tool_call)

                result_content = (
                    json.dumps(tool_result.result)
                    if not isinstance(tool_result.result, str)
                    else tool_result.result
                )
                if tool_result.error:
                    result_content = f"Error: {tool_result.error}"

                yield StreamEvent(
                    type=StreamEventType.TOOL_RESULT,
                    tool_name=tool_call.name,
                    tool_call_id=tool_call.id,
                    data=result_content,
                )

                messages.append(
                    Message(
                        role=Role.TOOL,
                        content=result_content,
                        tool_call_id=tool_call.id,
                        name=tool_call.name,
                    )
                )

        yield StreamEvent(type=StreamEventType.DONE)

    def _build_messages(self, prompt: str) -> list[Message]:
        """Build the message list for an LLM call."""
        messages: list[Message] = []

        # System prompt
        if self.instructions:
            messages.append(Message(role=Role.SYSTEM, content=self.instructions))

        # History (if memory enabled)
        if self.memory and self._history:
            # Skip system message from history (we already added it)
            for msg in self._history:
                if msg.role != Role.SYSTEM:
                    messages.append(msg)

        # User prompt
        messages.append(Message(role=Role.USER, content=prompt))

        return messages

    def get_tool_definitions(self) -> list[ToolDefinition]:
        """Get JSON Schema definitions for all registered tools."""
        return [tool.definition for tool in self.tools]

    def clear_memory(self) -> None:
        """Clear the agent's conversation history."""
        self._history.clear()

    def __repr__(self) -> str:
        tool_names = [t.name for t in self.tools] if self.tools else []
        return (
            f"Agent(model={self.model!r}, "
            f"tools={tool_names}, "
            f"memory={self.memory})"
        )


class ToolExecutor:
    """Executes tool calls by routing to the appropriate provider.

    Maps tool names to their execution backends:
    - code_exec → DockerSandbox (local) or ScalixSandbox (cloud)
    - sql → SQLiteProvider (local) or ScalixDB (cloud)
    - web_search → httpx-based search
    - http → httpx request
    - Custom tools → their registered execute functions
    """

    def __init__(self, tools: list[Any], config: ScalixConfig) -> None:
        self._tools = {tool.name: tool for tool in tools}
        self._config = config
        self._sandbox: Any = None
        self._database: Any = None

    async def execute(self, tool_call: ToolCall) -> ToolCallResult:
        """Execute a tool call and return the result."""
        start_time = time.monotonic()

        tool = self._tools.get(tool_call.name)
        if tool is None:
            return ToolCallResult(
                tool_name=tool_call.name,
                arguments=tool_call.arguments,
                result=None,
                error=f"Unknown tool: {tool_call.name}",
            )

        try:
            if tool_call.name == "code_exec":
                result = await self._execute_code(tool_call.arguments)
            elif tool_call.name == "sql":
                result = await self._execute_sql(tool_call.arguments)
            elif tool_call.name == "web_search":
                result = await self._execute_web_search(tool_call.arguments)
            elif tool_call.name == "http":
                result = await self._execute_http(tool_call.arguments)
            elif tool.has_execute_fn:
                # Custom tool with registered function
                result = await tool.execute(**tool_call.arguments)
                return result
            else:
                result = f"Tool '{tool_call.name}' has no execution handler"

            duration = (time.monotonic() - start_time) * 1000
            return ToolCallResult(
                tool_name=tool_call.name,
                arguments=tool_call.arguments,
                result=result,
                duration_ms=duration,
            )
        except Exception as e:
            duration = (time.monotonic() - start_time) * 1000
            logger.error("Tool '%s' failed: %s", tool_call.name, str(e))
            return ToolCallResult(
                tool_name=tool_call.name,
                arguments=tool_call.arguments,
                result=None,
                error=str(e),
                duration_ms=duration,
            )

    async def _get_sandbox(self) -> Any:
        """Get sandbox provider (lazy init)."""
        if self._sandbox is None:
            if self._config.is_cloud_mode:
                from scalix.providers.cloud.sandbox import ScalixSandboxProvider

                self._sandbox = ScalixSandboxProvider(self._config)
            else:
                from scalix.providers.local.docker_sandbox import DockerSandbox

                self._sandbox = DockerSandbox()
        return self._sandbox

    async def _get_database(self) -> Any:
        """Get database provider (lazy init)."""
        if self._database is None:
            from scalix.providers.local.sqlite_store import SQLiteProvider

            self._database = SQLiteProvider()
        return self._database

    async def _execute_code(self, arguments: dict[str, Any]) -> Any:
        """Execute code via sandbox provider."""
        sandbox = await self._get_sandbox()
        code = arguments.get("code", "")
        runtime = arguments.get("runtime", "python")
        result = await sandbox.execute(code=code, runtime=runtime)
        output = result.stdout
        if result.stderr:
            output += f"\nSTDERR:\n{result.stderr}"
        if result.exit_code != 0:
            output += f"\n[Exit code: {result.exit_code}]"
        return output

    async def _execute_sql(self, arguments: dict[str, Any]) -> Any:
        """Execute SQL via database provider."""
        db = await self._get_database()
        query = arguments.get("query", "")
        rows = await db.query(query)
        return rows

    async def _execute_web_search(self, arguments: dict[str, Any]) -> Any:
        """Execute web search via Scalix API or configurable endpoint."""
        import httpx
        from ..config import get_config

        query = arguments.get("query", "")
        config = get_config()
        headers: dict[str, str] = {"User-Agent": "Scalix-SDK/0.1"}

        if config.search_base_url:
            search_url = config.search_base_url
        else:
            search_url = f"{config.base_url}/v1/search"
        if config.api_key:
            headers["Authorization"] = f"Bearer {config.api_key}"

        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.get(
                search_url,
                params={"q": query},
                headers=headers,
            )
            return resp.text[:5000]

    async def _execute_http(self, arguments: dict[str, Any]) -> Any:
        """Execute HTTP request."""
        import httpx

        url = arguments.get("url", "")
        method = arguments.get("method", "GET").upper()
        body = arguments.get("body")

        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.request(method, url, json=body)
            return {"status": resp.status_code, "body": resp.text[:5000]}
